### ISSUE TITLE
How to effectively model different type of (n + 1) 

### ISSUE THREAD

**Issue by luminescent** [sangria-graphql/sangria#106]
Hi 

I am trying to use sangria and model the following structure: 

```
items [{
   id
   details {
     title
   }
 }{
   id
   details {
     title
   }
 }
...
]
```
(Not sure if my syntax above is clear, basically I want to retrieve a list of items, each item has an id and details)

This should be built on top of an existing REST API microservices framework, and the calls are as follows: 
*  one call to one service retrieves the items ids; let's call it the ids service  
*  different calls to another service retrieve the item details; this one is the details service 

I am not sure how to model this efficiently in sangria. I would like to attain the followings:
* if details are not requested, so one asks only for  `{items {id}}`, I don't want to make any calls to the details service, just make one call to the ids service to get the ids; 
* if details are requested (`{items {id, details {title}}}`), the calls to the services should be as follows: one call to the ids service, to get the ids, followed by calls fired up in parallel to the details service, to get the details; getting all the details sequentially would take too long.  

Any help would be greatly appreciated!

**Comment by OlegIlyenko** [comment]
You can always return a `Future` from a `resolve` function. Is will allow you to fetch the item details in parallel:

```scala
val ItemType =
  ObjectType("Item",
    fields[ItemService, Int](
      Field("id", StringType, resolve = _.value),
      Field("details", OptionType(ItemDetailsType),
        resolve = c ⇒
          c.ctx.getItemDetailAsFuture(id = c.value))
    ))
```

Sangria also supports a deferred values. It can be helpful if you are able to fetch multiple item details in one request. This will allow you to transform N+1 requests into 2 very efficient requests to external services. One very similar use-case is described here: https://github.com/sangria-graphql/sangria/issues/63

Here are also some pointers to the relevant parts of the documentation:

* http://sangria-graphql.org/learn/#deferred-values-and-resolver
* http://sangria-graphql.org/learn/#actions

Hope this helps. If not, please let me know :)

**Comment by luminescent** [comment]
Thank you for your quick reply, it was really helpful. 

The details service does not allow batched requests at this time so I have implemented, point by point, what you described in the code snippet, which is quite useful as the calls for details run in parallel.
 
More so, the service will allow retrieving details for multiple ids in one call, sometime in the near future, so #63 will come in handy as well. 

I am closing the issue as my question has been answered and thank you for Sangria, I am positively surprised by how little code I have to put in to have a fully functional application.
